### PR TITLE
Use Microsoft.CodeAnalysis.PublicApiAnalyzers

### PR DESCRIPTION
**Pull request by dougbu**
- #4259 1/2
- followup 2/3 for 5266918ed2be
- includes baselines for 16 MVC projects
    - will automated further additions in another PR
- suppress warnings that may cause back-compat problems if fixed

nit: sort `@(LatestPackageReference)` a bit better

### CODE REVIEW

File: eng/targets/CSharp.Common.targets [hunk 1]
```diff
@@ -6,6 +6,30 @@
     <TargetFrameworkIdentifier>.NETFramework</TargetFrameworkIdentifier>
   </PropertyGroup>
 
+  <!-- Ensure API changes show up clearly in PRs. -->
+  <PropertyGroup>
+    <_TFMDirectory>$(TargetFramework)</_TFMDirectory>
+    <_TFMDirectory Condition=" '$(TargetFramework)' == '$(DefaultNetCoreTargetFramework)' ">netcoreapp</_TFMDirectory>
+
+    <!-- Public members should not use oblivious types. Not done with all nullable annotations. -->
+    <NoWarn>$(NoWarn);RS0041</NoWarn>
```
Comment (dougbu) [inline eng/targets/CSharp.Common.targets:15]: @pranavkm where are we on our nullable annotations journey❔

File: src/Mvc/Mvc.Core/src/Routing/ControllerLinkGeneratorExtensions.cs [hunk 2]
```diff
@@ -36,7 +36,9 @@ public static class ControllerLinkGeneratorExtensions
         /// names from <c>RouteOptions</c>.
         /// </param>
         /// <returns>A URI with an absolute path, or <c>null</c> if a URI cannot be created.</returns>
+#pragma warning disable RS0026 // Do not add multiple public overloads with optional parameters
         public static string GetPathByAction(
+#pragma warning restore RS0026 // Do not add multiple public overloads with optional parameters
```
Comment (dougbu) [inline src/Mvc/Mvc.Core/src/Routing/ControllerLinkGeneratorExtensions.cs:41]: @ajaybhargavb @JamesNK I realize this and src/Mvc/Mvc.Core/src/Routing/PageLinkGeneratorExtensions.cs aren't particularly new. But, can this problem be addressed and not suppressed without breaking back-compat❔ If not, do you suggest we take the hit now❔

Comment (JamesNK) [reply src/Mvc/Mvc.Core/src/Routing/ControllerLinkGeneratorExtensions.cs:41]: The only way to resolve the warning is changing method signatures which is a breaking change.

I don't know of any issues caused by the overloads. Suppressing is fine.

@javiercn

File: eng/targets/CSharp.Common.targets (hunk 1 above)
Comment (pranavkm) [reply eng/targets/CSharp.Common.targets:15]: Still some ways to go. But you could help: https://github.com/dotnet/aspnetcore/pull/23475

https://giphy.com/gifs/2wG65JG3UAKpW/html5

**Review by JunTaoLuo — APPROVED** [review]
LGTM

File: src/Mvc/Mvc.Core/src/Routing/ControllerLinkGeneratorExtensions.cs (hunk 2 above)
Comment (sharwell) [reply src/Mvc/Mvc.Core/src/Routing/ControllerLinkGeneratorExtensions.cs:41]: I would recommend using attribute-style suppressions for methods, with the Justification set to "Required to maintain compatibility".

**Review by Pilchie — APPROVED** [review]

**Comment by Pilchie** [comment]
I suggest that once we fork for RC1, we move all of those items to the `.Shipped` files so that we can track any changes that happen.

**Comment by dougbu** [comment]
> I suggest that once we fork for RC1, we move all of those items to the `.Shipped` files so that we can track any changes that happen.

Agreed. Should I disable any other warnings from https://github.com/dotnet/roslyn-analyzers/blob/master/src/PublicApiAnalyzers/Microsoft.CodeAnalysis.PublicApiAnalyzers.md in the meantime❔

In my next PR, I need to
1. add a small document covering planned use of these files and link to the documentation of the analyzer package
    - I'll include a note about RC1 plans here
2. add lots o' `PublicApi.*.txt files`
3. automate this infrastructure more e.g. provide a way to automatically add missing `PublicApi.*.txt files` and update them (if possible) from the command line

**Comment by Pilchie** [comment]
I don't think we want to disable any of the other warnings.

File: src/Mvc/Mvc.Core/src/Routing/ControllerLinkGeneratorExtensions.cs (hunk 2 above)
Comment (dougbu) [reply src/Mvc/Mvc.Core/src/Routing/ControllerLinkGeneratorExtensions.cs:41]: That seems to work but VS tends to crash or ignore me when using attribute suppressions for this issue. I eventually got it to fix one instance and copied my brand new attribute everywhere else. Something may be busted in this area.

**Comment by dougbu** [comment]
Issue about the second item on my list is #24347. Please help if you have time.

**Comment by dougbu** [comment]
A new version of Microsoft.CodeAnalysis.PublicApiAnalyzers containing @mavasani's dotnet/roslyn-analyzers#3916 improvement will address the last item on my list. We'll pick that up when it's ready.

Note: The new package will make missing `PublicApi.*.txt` files fail our builds because we treat warnings as errors. Anything left in #24347 will need to be done in one fell swoop unless we make use of the package conditional on the files existing.

Good news is we can delete https://github.com/dotnet/aspnetcore/blob/0437117cfb5bc91a34878fb962076bb05f4b8273/eng/targets/CSharp.Common.targets#L23-L30 when using the new package.

**Comment by mavasani** [comment]
> Hmm, probably need to keep the $(_TFMDirectory) items for projects with slightly different public APIs per TFM.

Yes, that sounds correct.

**Pull request merged**